server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,522 @@
+        """
+        :param repo: Pulp managed Yum repository
+        :type  repo: pulp.plugins.model.Repository
+        :param publish_conduit: Conduit providing access to relative Pulp functionality
+        :type  publish_conduit: pulp.plugins.conduits.repo_publish.RepoPublishConduit
+        :param config: Pulp configuration for the distributor
+        :type  config: pulp.plugins.config.PluginCallConfiguration
+        """
+
+        super(Publisher, self).__init__("Repository publish", repo,
+                                        publish_conduit, config,
+                                        distributor_type=distributor_type,
+                                        working_dir="")
+
+        distributor = Distributor.objects.get_or_404(repo_id=self.repo.id,
+                                                     distributor_id=publish_conduit.distributor_id)
+        self.last_published = distributor["last_publish"]
+        self.last_deleted = repo.last_unit_removed
+        self.repo_type = repo.notes["_repo-type"]
+        self.predistributor = self._get_predistributor()
+        if self.last_published and not isinstance(self.last_published, basestring):
+            string_date = dateutils.format_iso8601_datetime(self.last_published)
+        elif self.last_published:
+            string_date = self.last_published
+        else:
+            string_date = None
+        if self.predistributor:
+            search_params = {'repo_id': repo.id,
+                             'distributor_id': self.predistributor["id"],
+                             'started': {"$gte": string_date}}
+            self.predist_history = RepoPublishResult.get_collection().find(search_params)
+        else:
+            self.predist_history = []
+
+        self.remote_path = self.get_remote_repo_path()
+
+        if not self.is_fastforward():
+            date_filter = None
+        else:
+            range_criteria = {START_DATE_KEYWORD: self.last_published,
+                              END_DATE_KEYWORD: self.predistributor["last_publish"]}
+            date_filter = self.create_date_range_filter(range_criteria)
+
+        self.extra_path = self._get_extra_path()
+        self.original_lists = {}
+        self.link_lists = {}
+
+        self._add_necesary_steps(date_filter=date_filter, config=config)
+
+    def is_fastforward(self):
+        """
+        This method checks whether this publish should be a fastforward publish.
+
+        :return: Whether or not this publish should be in fast forward mode
+        :rtype: bool
+        """
+        skip_fast_forward = False
+        for entry in self.predist_history:
+            str_skip_ff = entry.get("distributor_config", {}).get("skip_fast_forward", False)
+            skip_fast_forward |= misc.strbool2bool(str_skip_ff)
+            if entry.get("result", "error") == "error":
+                skip_fast_forward = True
+
+        if self.last_published and isinstance(self.last_published, basestring):
```

Comment:
Rather than continuing to check what type this is every time it gets used, let's do it just once in ``__init__``, or not at all if the model takes care of it.